toolchain/diagnostics/diagnostic_emitter.h

```diff
@@ -168,20 +243,23 @@ class DiagnosticEmitter {
   // When passing arguments, they may be buffered. As a consequence, lifetimes
   // may outlive the `Emit` call.
   template <typename... Args>
-  void Emit(LocationT location,
+  auto Emit(LocationT location,
             const Internal::DiagnosticBase<Args...>& diagnostic_base,
             // Disable type deduction based on `args`; the type of
             // `diagnostic_base` determines the diagnostic's parameter types.
-            typename std::common_type_t<Args>... args) {
-    consumer_->HandleDiagnostic({
-        .kind = diagnostic_base.Kind,
-        .level = diagnostic_base.Level,
-        .location = translator_->GetLocation(location),
-        .format = diagnostic_base.Format,
-        .format_args = {std::move(args)...},
-        .format_fn = [&diagnostic_base](const Diagnostic& diagnostic)
-            -> std::string { return diagnostic_base.FormatFn(diagnostic); },
-    });
+            typename std::common_type_t<Args>... args) -> void {
+    DiagnosticBuilder(this, location, diagnostic_base, std::move(args)...)
+        .Emit();
+  }
+
+  template <typename... Args>
+  auto Build(LocationT location,
```

Comment:
Seems good to have some documentation here, especially maybe an example since this is a fluent-API where that might be more important than other kinds of APIs?

Reply:
Done